{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "01334395-74cd-4de4-87fe-487f20f43823",
   "metadata": {},
   "source": [
    "# DDSP Training that works\n",
    "\n",
    "\n",
    "## Prerequisites\n",
    "\n",
    "1. Macbook with Apple Silicon\n",
    "2. Python 3.8 (using `uv` package manager).\n",
    "   1. Create a virtual evironment before running this notebook by `uv venv --python 3.8`, which creates a `.venv` directory. You can activate the venv by `source .venv/bin/activate`.\n",
    "   2. Install jupyter notebook by `uv pip install jupyter notebook`.  \n",
    "3. Local-install `ddsp`\n",
    "   1. `git clone https://github.com/magenta/ddsp.git`, make sure the commit is `a2d0517`.\n",
    "   2. Modify `install_requires` in `setup.py` by removing all version restriction.\n",
    "   3. Install locally by `uv pip install -e ./ddsp`\n",
    "4. Install `tensorflow-metal` by `uv pip install tensorflow-metal` to GPU usage in tensorflow."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6742d8bd-a9e0-4399-bc2e-0b6684840154",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Dependency List\n",
    "```\n",
    "absl-py==1.4.0\n",
    "anyio==4.5.2\n",
    "apache-beam==2.60.0\n",
    "appnope==0.1.4\n",
    "argon2-cffi==23.1.0\n",
    "argon2-cffi-bindings==21.2.0\n",
    "array-record==0.4.0\n",
    "arrow==1.3.0\n",
    "asttokens==3.0.0\n",
    "astunparse==1.6.3\n",
    "async-lru==2.0.4\n",
    "async-timeout==5.0.1\n",
    "attrs==25.3.0\n",
    "audioread==3.0.1\n",
    "babel==2.17.0\n",
    "backcall==0.2.0\n",
    "beautifulsoup4==4.13.4\n",
    "bleach==6.1.0\n",
    "bokeh==3.1.1\n",
    "cached-property==2.0.1\n",
    "cachetools==5.5.2\n",
    "certifi==2025.4.26\n",
    "cffi==1.17.1\n",
    "charset-normalizer==3.4.2\n",
    "chex==0.1.7\n",
    "click==8.1.8\n",
    "cloudml-hypertune==0.1.0.dev6\n",
    "cloudpickle==2.2.1\n",
    "comm==0.2.2\n",
    "contourpy==1.1.1\n",
    "crcmod==1.7\n",
    "crepe==0.0.16\n",
    "cycler==0.12.1\n",
    "debugpy==1.8.14\n",
    "decorator==5.2.1\n",
    "defusedxml==0.7.1\n",
    "dill==0.3.1.1\n",
    "dm-tree==0.1.8\n",
    "dnspython==2.6.1\n",
    "docopt==0.6.2\n",
    "etils==1.3.0\n",
    "exceptiongroup==1.2.2\n",
    "executing==2.2.0\n",
    "fastavro==1.9.7\n",
    "fasteners==0.19\n",
    "fastjsonschema==2.21.1\n",
    "flatbuffers==25.2.10\n",
    "flax==0.7.2\n",
    "fonttools==4.57.0\n",
    "fqdn==1.5.1\n",
    "future==1.0.0\n",
    "gast==0.4.0\n",
    "gin-config==0.5.0\n",
    "google-api-core==2.24.2\n",
    "google-auth==2.40.1\n",
    "google-auth-oauthlib==1.0.0\n",
    "google-cloud-core==2.4.3\n",
    "google-cloud-storage==3.1.0\n",
    "google-crc32c==1.5.0\n",
    "google-pasta==0.2.0\n",
    "google-resumable-media==2.7.2\n",
    "googleapis-common-protos==1.70.0\n",
    "grpcio==1.65.5\n",
    "h11==0.16.0\n",
    "h5py==3.11.0\n",
    "hdfs==2.7.3\n",
    "hmmlearn==0.3.3\n",
    "httpcore==1.0.9\n",
    "httplib2==0.22.0\n",
    "httpx==0.28.1\n",
    "idna==3.10\n",
    "imageio==2.35.1\n",
    "importlib-metadata==8.5.0\n",
    "importlib-resources==6.4.5\n",
    "intervaltree==3.1.0\n",
    "ipyfilechooser==0.6.0\n",
    "ipykernel==6.29.5\n",
    "ipython==8.12.3\n",
    "ipywidgets==8.1.7\n",
    "isoduration==20.11.0\n",
    "jax==0.4.13\n",
    "jaxlib==0.4.13\n",
    "jedi==0.19.2\n",
    "jinja2==3.1.6\n",
    "joblib==1.4.2\n",
    "json5==0.12.0\n",
    "jsonpickle==3.4.2\n",
    "jsonpointer==3.0.0\n",
    "jsonschema==4.23.0\n",
    "jsonschema-specifications==2023.12.1\n",
    "jupyter==1.1.1\n",
    "jupyter-client==8.6.3\n",
    "jupyter-console==6.6.3\n",
    "jupyter-core==5.7.2\n",
    "jupyter-events==0.10.0\n",
    "jupyter-lsp==2.2.5\n",
    "jupyter-server==2.14.2\n",
    "jupyter-server-terminals==0.5.3\n",
    "jupyterlab==4.3.7\n",
    "jupyterlab-pygments==0.3.0\n",
    "jupyterlab-server==2.27.3\n",
    "jupyterlab-widgets==3.0.15\n",
    "keras==2.13.1\n",
    "kiwisolver==1.4.7\n",
    "lazy-loader==0.4\n",
    "libclang==18.1.1\n",
    "librosa==0.11.0\n",
    "llvmlite==0.41.1\n",
    "markdown==3.7\n",
    "markdown-it-py==3.0.0\n",
    "markupsafe==2.1.5\n",
    "matplotlib==3.7.5\n",
    "matplotlib-inline==0.1.7\n",
    "mdurl==0.1.2\n",
    "mido==1.3.3\n",
    "mir-eval==0.8.2\n",
    "mistune==3.1.3\n",
    "ml-dtypes==0.2.0\n",
    "msgpack==1.1.0\n",
    "nbclient==0.10.1\n",
    "nbconvert==7.16.6\n",
    "nbformat==5.10.4\n",
    "nest-asyncio==1.6.0\n",
    "note-seq==0.0.3\n",
    "notebook==7.3.3\n",
    "notebook-shim==0.2.4\n",
    "numba==0.58.1\n",
    "numpy==1.24.3\n",
    "oauthlib==3.2.2\n",
    "objsize==0.7.1\n",
    "opt-einsum==3.4.0\n",
    "optax==0.1.8\n",
    "orbax-checkpoint==0.2.3\n",
    "orjson==3.10.15\n",
    "overrides==7.7.0\n",
    "packaging==23.2\n",
    "pandas==2.0.3\n",
    "pandocfilters==1.5.1\n",
    "parso==0.8.4\n",
    "pexpect==4.9.0\n",
    "pickleshare==0.7.5\n",
    "pillow==10.4.0\n",
    "pkgutil-resolve-name==1.3.10\n",
    "platformdirs==4.3.6\n",
    "pooch==1.8.2\n",
    "pretty-midi==0.2.10\n",
    "prometheus-client==0.21.1\n",
    "promise==2.3\n",
    "prompt-toolkit==3.0.51\n",
    "proto-plus==1.26.1\n",
    "protobuf==3.20.3\n",
    "psutil==7.0.0\n",
    "ptyprocess==0.7.0\n",
    "pure-eval==0.2.3\n",
    "pyarrow==16.1.0\n",
    "pyarrow-hotfix==0.7\n",
    "pyasn1==0.6.1\n",
    "pyasn1-modules==0.4.2\n",
    "pybind11==2.13.6\n",
    "pycparser==2.22\n",
    "pydot==1.4.2\n",
    "pydub==0.25.1\n",
    "pygments==2.19.1\n",
    "pyjwt==2.9.0\n",
    "pymongo==4.10.1\n",
    "pyparsing==3.1.4\n",
    "python-dateutil==2.9.0.post0\n",
    "python-json-logger==3.3.0\n",
    "pytz==2025.2\n",
    "pyyaml==6.0.2\n",
    "pyzmq==26.4.0\n",
    "redis==5.3.0\n",
    "referencing==0.35.1\n",
    "regex==2024.11.6\n",
    "requests==2.32.3\n",
    "requests-oauthlib==2.0.0\n",
    "resampy==0.4.3\n",
    "rfc3339-validator==0.1.4\n",
    "rfc3986-validator==0.1.1\n",
    "rich==14.0.0\n",
    "rpds-py==0.20.1\n",
    "rsa==4.9.1\n",
    "scikit-learn==1.3.2\n",
    "scipy==1.10.1\n",
    "send2trash==1.8.3\n",
    "setuptools==75.3.2\n",
    "six==1.17.0\n",
    "sniffio==1.3.1\n",
    "sortedcontainers==2.4.0\n",
    "soundfile==0.13.1\n",
    "soupsieve==2.7\n",
    "soxr==0.3.7\n",
    "stack-data==0.6.3\n",
    "tensorboard==2.13.0\n",
    "tensorboard-data-server==0.7.2\n",
    "tensorflow==2.13.0\n",
    "tensorflow-datasets==4.9.2\n",
    "tensorflow-decision-forests==1.5.0\n",
    "tensorflow-estimator==2.13.0\n",
    "tensorflow-hub==0.16.1\n",
    "tensorflow-macos==2.13.0\n",
    "tensorflow-metadata==1.14.0\n",
    "tensorflow-metal==1.0.1\n",
    "tensorflow-probability==0.21.0\n",
    "tensorflowjs==4.22.0\n",
    "tensorstore==0.1.45\n",
    "termcolor==2.4.0\n",
    "terminado==0.18.1\n",
    "tf-keras==2.15.0\n",
    "tflite-support==0.1.0a1\n",
    "threadpoolctl==3.5.0\n",
    "tinycss2==1.2.1\n",
    "toml==0.10.2\n",
    "tomli==2.2.1\n",
    "toolz==1.0.0\n",
    "tornado==6.4.2\n",
    "tqdm==4.67.1\n",
    "traitlets==5.14.3\n",
    "types-python-dateutil==2.9.0.20241206\n",
    "typing-extensions==4.5.0\n",
    "tzdata==2025.2\n",
    "uri-template==1.3.0\n",
    "urllib3==2.2.3\n",
    "wcwidth==0.2.13\n",
    "webcolors==24.8.0\n",
    "webencodings==0.5.1\n",
    "websocket-client==1.8.0\n",
    "werkzeug==3.0.6\n",
    "wheel==0.45.1\n",
    "widgetsnbextension==4.0.14\n",
    "wrapt==1.17.2\n",
    "wurlitzer==3.1.1\n",
    "xyzservices==2025.4.0\n",
    "zipp==3.20.2\n",
    "zstandard==0.23.0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45d12072-2e98-48b5-b5d9-2a063c365359",
   "metadata": {},
   "source": [
    "## Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5166039d-a46f-4b9e-a3d1-1e95212abe4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Name = \"gayageum\"  # model name\n",
    "Training_Steps = 30000  # 30k or higher\n",
    "Ignore_Previous = False # resume\n",
    "Sample_Dir = './ddsp_train/gayageum_samples_all'  # audio sample directory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fea64ec6-a2d7-4c72-a73c-57b9506e8dd1",
   "metadata": {},
   "source": [
    "## Extra Dependency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91cf9cf3-e045-4657-a759-ea6eb593c630",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install ipyfilechooser==0.6.0 &> /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bf8e9d-f065-4bb1-9929-3cca3de15578",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a1706cb1-0c2c-4adb-bf6d-a69bd27638a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import datetime\n",
    "import glob\n",
    "import os\n",
    "import shutil\n",
    "import time\n",
    "import IPython\n",
    "import json\n",
    "import subprocess\n",
    "\n",
    "# from google.colab import drive\n",
    "import tensorflow as tf\n",
    "\n",
    "from ipyfilechooser import FileChooser\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3a59d6-ba4b-4b4b-a027-2b1fdfa1b02f",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "53f2f20f-ab4b-4930-a8a6-f5ada74e50ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def log_event(event_name, event_details):\n",
    "  \"\"\"Log event with name and details dictionary.\"\"\"\n",
    "  details_json = json.dumps(event_details)\n",
    "  js_string = \"gtag('event', '%s', %s);\" % (event_name, details_json)\n",
    "  IPython.display.display(IPython.display.Javascript(js_string))\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# Functions\n",
    "# ------------------------------------------------------------------------------\n",
    "def directory_has_files(target_dir):\n",
    "  n_files = len(glob.glob(os.path.join(target_dir, '*')))\n",
    "  return n_files > 0\n",
    "\n",
    "\n",
    "def get_audio_files(drive_dir, audio_dir):\n",
    "  if drive_dir:\n",
    "    mp3_files = glob.glob(os.path.join(drive_dir, '*.mp3'))\n",
    "    wav_files = glob.glob(os.path.join(drive_dir, '*.wav'))\n",
    "    audio_paths = mp3_files + wav_files\n",
    "    if len(audio_paths) < 1:\n",
    "      raise FileNotFoundError(\"Sorry, it seems that there aren't any MP3 or \"\n",
    "                              f\"WAV files in your folder ({drive_dir}). Try \"\n",
    "                              \"running again and choose a different folder.\")\n",
    "  else:\n",
    "    audio_paths, _ = colab_utils.upload()\n",
    "\n",
    "  # Copy Audio.\n",
    "  print('Copying audio to colab for training...')\n",
    "  for src in audio_paths:\n",
    "    target = os.path.join(audio_dir, \n",
    "                          os.path.basename(src).replace(' ', '_'))\n",
    "    print('Copying {} to {}'.format(src, target))\n",
    "    shutil.copy(src, target)\n",
    "    # !cp $src $target\n",
    "\n",
    "\n",
    "def prepare_dataset(audio_dir, \n",
    "                    data_dir,\n",
    "                    sample_rate=16000, \n",
    "                    frame_rate=50, \n",
    "                    example_secs=4.0, \n",
    "                    hop_secs=1.0, \n",
    "                    viterbi=True, \n",
    "                    center=True):\n",
    "  if directory_has_files(data_dir):\n",
    "    print(f'Dataset already exists in `{data_dir}`')\n",
    "    return\n",
    "  else:\n",
    "    # Otherwise prepare new dataset locally.\n",
    "    print(f'Preparing new dataset from `{audio_dir}`')\n",
    "\n",
    "    print()\n",
    "    print('Creating dataset...')\n",
    "    print('This usually takes around 2-3 minutes for each minute of audio')\n",
    "    print('(10 minutes of training audio -> 20-30 minutes)')\n",
    "\n",
    "    audio_filepattern = os.path.join(audio_dir, '*')\n",
    "    audio_fp_str = f'\"{audio_filepattern}\"'    \n",
    "    tfrecord_path_str = f'\"{data_dir}/train.tfrecord\"'\n",
    "\n",
    "    !ddsp_prepare_tfrecord \\\n",
    "    --input_audio_filepatterns=$audio_fp_str \\\n",
    "    --output_tfrecord_path=$tfrecord_path_str \\\n",
    "    --num_shards=10 \\\n",
    "    --sample_rate=$sample_rate \\\n",
    "    --frame_rate=$frame_rate \\\n",
    "    --example_secs=$example_secs \\\n",
    "    --hop_secs=$hop_secs \\\n",
    "    --viterbi=$viterbi \\\n",
    "    --center=$center\n",
    "    #  &> /dev/null\n",
    "\n",
    "\n",
    "def train(model_dir, data_dir, steps=30000):\n",
    "  file_pattern = os.path.join(data_dir, 'train.tfrecord*')\n",
    "  fp_str = f\"TFRecordProvider.file_pattern='{file_pattern}'\"\n",
    "  !ddsp_run \\\n",
    "  --mode=train \\\n",
    "  --save_dir=\"$model_dir\" \\\n",
    "  --gin_file=models/vst/vst.gin \\\n",
    "  --gin_file=datasets/tfrecord.gin \\\n",
    "  --gin_param=\"$fp_str\" \\\n",
    "  --gin_param=\"TFRecordProvider.centered=True\" \\\n",
    "  --gin_param=\"TFRecordProvider.frame_rate=50\" \\\n",
    "  --gin_param=\"batch_size=16\" \\\n",
    "  --gin_param=\"train_util.train.num_steps=$steps\" \\\n",
    "  --gin_param=\"train_util.train.steps_per_save=300\" \\\n",
    "  --gin_param=\"trainers.Trainer.checkpoints_to_keep=3\"\n",
    "\n",
    "  # --gin_param=\"train.data_provider=@ExperimentalDataProvider()\" \\\n",
    "  # --gin_param=\"ExperimentalRecordProvider.data_dir='$data_dir'\" \\\n",
    "  # --gin_param=\"ExperimentalRecordProvider.sample_rate=16000\" \\\n",
    "  # --gin_param=\"ExperimentalRecordProvider.frame_rate=50\" \\\n",
    "\n",
    "\n",
    "def reset_state(data_dir, audio_dir, model_dir):\n",
    "  model_dir_str = f'\"{model_dir}\"'\n",
    "  if tf.io.gfile.exists(data_dir):\n",
    "    !rm -r $data_dir\n",
    "    !rm -r $audio_dir\n",
    "  !mkdir -p $data_dir\n",
    "  !mkdir -p $audio_dir\n",
    "  !mkdir -p $model_dir_str\n",
    "\n",
    "\n",
    "def export_and_download(model_dir, model_name=Name):\n",
    "  export_path = os.path.join(model_dir, model_name)\n",
    "\n",
    "  model_dir_str=f'\"{model_dir}\"'\n",
    "  export_path_str=f'\"{export_path}\"'\n",
    "  \n",
    "  !ddsp_export \\\n",
    "  --name=$model_name \\\n",
    "  --model_path=$model_dir_str \\\n",
    "  --save_dir=$export_path_str \\\n",
    "  --inference_model=vst_stateless_predict_controls \\\n",
    "  --tflite \\\n",
    "  --notfjs\n",
    "\n",
    "  # Zip the whole directory.\n",
    "  zip_fname = f'{model_name}.zip'\n",
    "  zip_fp = os.path.join(model_dir, zip_fname)\n",
    "  print(f'Export complete! Zipping {export_path} to {zip_fp}')\n",
    "  !cd $model_dir_str && zip -r $zip_fname ./$model_name\n",
    "\n",
    "  # Download.\n",
    "  print(f'Zipping Complete! Downloading... {zip_fname}')\n",
    "  print(f'You can also find your model at {export_path}')\n",
    "  # colab_utils.download(zip_fp)\n",
    "\n",
    "\n",
    "def get_model_dir(base_dir):\n",
    "  base_str = 'ddsp-training'\n",
    "  dirs = tf.io.gfile.glob(os.path.join(base_dir, f'{base_str}-*'))\n",
    "  if dirs and not Ignore_Previous:\n",
    "    model_dir = dirs[-1]  # Sorted, so last is most recent.\n",
    "  else:\n",
    "    now = datetime.datetime.now().strftime('%Y-%m-%d-%H%M')\n",
    "    model_dir = os.path.join(base_dir, f'{base_str}-{now}')\n",
    "  return model_dir\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89065bf-ee24-4fbe-9f8c-68316f57dc9a",
   "metadata": {},
   "source": [
    "## Runner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e736b9cd-e3c5-4c91-958d-ada8588ce1c9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/javascript": "gtag('event', 'runStarted', {});",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "gtag('event', 'gpuType', {\"event_category\": \"mac\"});",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "gtag('event', 'trainingLocally', {});",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "gtag('event', 'runTrainingStarted', {});",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training...\n",
      "I0518 15:02:09.992473 8667810944 ddsp_run.py:179] Restore Dir: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917\n",
      "I0518 15:02:09.992578 8667810944 ddsp_run.py:180] Save Dir: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917\n",
      "I0518 15:02:09.992632 8667810944 resource_reader.py:50] system_path_file_exists:optimization/base.gin\n",
      "E0518 15:02:09.992692 8667810944 resource_reader.py:55] Path not found: optimization/base.gin\n",
      "I0518 15:02:09.993370 8667810944 resource_reader.py:50] system_path_file_exists:eval/basic.gin\n",
      "E0518 15:02:09.993418 8667810944 resource_reader.py:55] Path not found: eval/basic.gin\n",
      "I0518 15:02:09.994170 8667810944 ddsp_run.py:148] Using operative config: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/operative_config-9000.gin\n",
      "I0518 15:02:09.996337 8667810944 resource_reader.py:50] system_path_file_exists:models/vst/vst.gin\n",
      "E0518 15:02:09.996389 8667810944 resource_reader.py:55] Path not found: models/vst/vst.gin\n",
      "I0518 15:02:09.998178 8667810944 resource_reader.py:50] system_path_file_exists:datasets/tfrecord.gin\n",
      "E0518 15:02:09.998245 8667810944 resource_reader.py:55] Path not found: datasets/tfrecord.gin\n",
      "I0518 15:02:09.998545 8667810944 resource_reader.py:50] system_path_file_exists:datasets/base.gin\n",
      "E0518 15:02:09.998586 8667810944 resource_reader.py:55] Path not found: datasets/base.gin\n",
      "I0518 15:02:10.000976 8667810944 ddsp_run.py:184] Operative Gin Config:\n",
      "import ddsp\n",
      "import ddsp.training as ddsp2\n",
      "\n",
      "# Macros:\n",
      "# ==============================================================================\n",
      "batch_size = 16\n",
      "evaluators = [@BasicEvaluator]\n",
      "frame_rate = 50\n",
      "frame_size = 1024\n",
      "learning_rate = 0.0003\n",
      "n_samples = 64320\n",
      "sample_rate = 16000\n",
      "\n",
      "# Parameters for processors.Add:\n",
      "# ==============================================================================\n",
      "processors.Add.name = 'add'\n",
      "\n",
      "# Parameters for Autoencoder:\n",
      "# ==============================================================================\n",
      "Autoencoder.decoder = @decoders.RnnFcDecoder()\n",
      "Autoencoder.encoder = None\n",
      "Autoencoder.losses = [@losses.SpectralLoss()]\n",
      "Autoencoder.preprocessor = @preprocessing.OnlineF0PowerPreprocessor()\n",
      "Autoencoder.processor_group = @processors.ProcessorGroup()\n",
      "\n",
      "# Parameters for Crop:\n",
      "# ==============================================================================\n",
      "Crop.crop_location = 'back'\n",
      "Crop.frame_size = 320\n",
      "Crop.name = 'crop'\n",
      "\n",
      "# Parameters for evaluate:\n",
      "# ==============================================================================\n",
      "evaluate.batch_size = 32\n",
      "evaluate.data_provider = @data.TFRecordProvider()\n",
      "evaluate.evaluator_classes = %evaluators\n",
      "evaluate.num_batches = 5\n",
      "\n",
      "# Parameters for exp_sigmoid:\n",
      "# ==============================================================================\n",
      "exp_sigmoid.exponent = 10.0\n",
      "exp_sigmoid.max_value = 2.0\n",
      "exp_sigmoid.threshold = 1e-07\n",
      "\n",
      "# Parameters for FilteredNoise:\n",
      "# ==============================================================================\n",
      "FilteredNoise.initial_bias = -5.0\n",
      "FilteredNoise.n_samples = %n_samples\n",
      "FilteredNoise.name = 'filtered_noise'\n",
      "FilteredNoise.scale_fn = @core.exp_sigmoid\n",
      "FilteredNoise.window_size = 0\n",
      "\n",
      "# Parameters for FilteredNoiseReverb:\n",
      "# ==============================================================================\n",
      "FilteredNoiseReverb.add_dry = True\n",
      "FilteredNoiseReverb.initial_bias = -3.0\n",
      "FilteredNoiseReverb.n_filter_banks = 32\n",
      "FilteredNoiseReverb.n_frames = 500\n",
      "FilteredNoiseReverb.name = 'reverb'\n",
      "FilteredNoiseReverb.reverb_length = 24000\n",
      "FilteredNoiseReverb.trainable = True\n",
      "FilteredNoiseReverb.window_size = 257\n",
      "\n",
      "# Parameters for get_model:\n",
      "# ==============================================================================\n",
      "get_model.model = @models.Autoencoder()\n",
      "\n",
      "# Parameters for Harmonic:\n",
      "# ==============================================================================\n",
      "Harmonic.amp_resample_method = 'linear'\n",
      "Harmonic.n_samples = %n_samples\n",
      "Harmonic.name = 'harmonic'\n",
      "Harmonic.normalize_below_nyquist = True\n",
      "Harmonic.sample_rate = %sample_rate\n",
      "Harmonic.scale_fn = @core.exp_sigmoid\n",
      "Harmonic.use_angular_cumsum = False\n",
      "\n",
      "# Parameters for OnlineF0PowerPreprocessor:\n",
      "# ==============================================================================\n",
      "OnlineF0PowerPreprocessor.compute_f0 = False\n",
      "OnlineF0PowerPreprocessor.compute_power = True\n",
      "OnlineF0PowerPreprocessor.crepe_saved_model_path = None\n",
      "OnlineF0PowerPreprocessor.frame_rate = %frame_rate\n",
      "OnlineF0PowerPreprocessor.frame_size = %frame_size\n",
      "OnlineF0PowerPreprocessor.padding = 'center'\n",
      "OnlineF0PowerPreprocessor.viterbi = False\n",
      "\n",
      "# Parameters for oscillator_bank:\n",
      "# ==============================================================================\n",
      "oscillator_bank.sum_sinusoids = True\n",
      "\n",
      "# Parameters for ProcessorGroup:\n",
      "# ==============================================================================\n",
      "ProcessorGroup.dag = \\\n",
      "    [(@synths.Harmonic(), ['amps', 'harmonic_distribution', 'f0_hz']),\n",
      "     (@synths.FilteredNoise(), ['noise_magnitudes']),\n",
      "     (@processors.Add(), ['filtered_noise/signal', 'harmonic/signal']),\n",
      "     (@effects.FilteredNoiseReverb(), ['add/signal']),\n",
      "     (@processors.Crop(), ['reverb/signal'])]\n",
      "\n",
      "# Parameters for RnnFcDecoder:\n",
      "# ==============================================================================\n",
      "RnnFcDecoder.ch = 256\n",
      "RnnFcDecoder.input_keys = ('pw_scaled', 'f0_scaled')\n",
      "RnnFcDecoder.layers_per_stack = 1\n",
      "RnnFcDecoder.output_splits = \\\n",
      "    (('amps', 1), ('harmonic_distribution', 60), ('noise_magnitudes', 65))\n",
      "RnnFcDecoder.rnn_channels = 512\n",
      "RnnFcDecoder.rnn_type = 'gru'\n",
      "RnnFcDecoder.stateless = False\n",
      "\n",
      "# Parameters for run_dag:\n",
      "# ==============================================================================\n",
      "run_dag.verbose = False\n",
      "\n",
      "# Parameters for sample:\n",
      "# ==============================================================================\n",
      "sample.batch_size = 16\n",
      "sample.ckpt_delay_secs = 300\n",
      "sample.data_provider = @data.TFRecordProvider()\n",
      "sample.evaluator_classes = %evaluators\n",
      "sample.num_batches = 1\n",
      "\n",
      "# Parameters for SpectralLoss:\n",
      "# ==============================================================================\n",
      "SpectralLoss.cumsum_freq_weight = 0.0\n",
      "SpectralLoss.delta_freq_weight = 0.0\n",
      "SpectralLoss.delta_time_weight = 0.0\n",
      "SpectralLoss.fft_sizes = (2048, 1024, 512, 256, 128, 64)\n",
      "SpectralLoss.logmag_weight = 1.0\n",
      "SpectralLoss.loss_type = 'L1'\n",
      "SpectralLoss.loudness_weight = 0.0\n",
      "SpectralLoss.mag_weight = 1.0\n",
      "SpectralLoss.name = 'spectral_loss'\n",
      "\n",
      "# Parameters for TFRecordProvider:\n",
      "# ==============================================================================\n",
      "TFRecordProvider.centered = True\n",
      "TFRecordProvider.example_secs = 4\n",
      "TFRecordProvider.file_pattern = 'data/train.tfrecord*'\n",
      "TFRecordProvider.frame_rate = 50\n",
      "TFRecordProvider.sample_rate = 16000\n",
      "\n",
      "# Parameters for train:\n",
      "# ==============================================================================\n",
      "train.batch_size = %batch_size\n",
      "train.data_provider = @data.TFRecordProvider()\n",
      "train.num_steps = 30000\n",
      "train.steps_per_save = 300\n",
      "train.steps_per_summary = 300\n",
      "\n",
      "# Parameters for Trainer:\n",
      "# ==============================================================================\n",
      "Trainer.checkpoints_to_keep = 3\n",
      "Trainer.grad_clip_norm = 3.0\n",
      "Trainer.learning_rate = %learning_rate\n",
      "Trainer.lr_decay_rate = 0.98\n",
      "Trainer.lr_decay_steps = 10000\n",
      "Trainer.restore_keys = None\n",
      "\n",
      "I0518 15:02:10.001561 8667810944 train_util.py:76] Defaulting to MirroredStrategy\n",
      "INFO:tensorflow:Using MirroredStrategy with devices ('/job:localhost/replica:0/task:0/device:GPU:0',)\n",
      "I0518 15:02:10.003007 8667810944 mirrored_strategy.py:419] Using MirroredStrategy with devices ('/job:localhost/replica:0/task:0/device:GPU:0',)\n",
      "W0518 15:02:10.028928 8667810944 optimizer.py:70] At this time, the v2.11+ optimizer `tf.keras.optimizers.Adam` runs slowly on M1/M2 Macs, please use the legacy Keras optimizer instead, located at `tf.keras.optimizers.legacy.Adam`.\n",
      "I0518 15:02:10.132004 8667810944 trainers.py:141] Building the model...\n",
      "Model: \"autoencoder\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " online_f0_power_preprocess  multiple                  0         \n",
      " or (OnlineF0PowerPreproces                                      \n",
      " sor)                                                            \n",
      "                                                                 \n",
      " rnn_fc_decoder (RnnFcDecod  multiple                  1873278   \n",
      " er)                                                             \n",
      "                                                                 \n",
      " processor_group (Processor  multiple                  16000     \n",
      " Group)                                                          \n",
      "                                                                 \n",
      " spectral_loss (SpectralLos  multiple                  0         \n",
      " s)                                                              \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 1889278 (7.21 MB)\n",
      "Trainable params: 1889278 (7.21 MB)\n",
      "Non-trainable params: 0 (0.00 Byte)\n",
      "_________________________________________________________________\n",
      "I0518 15:02:13.773509 8667810944 trainers.py:94] Restoring from checkpoint...\n",
      "I0518 15:02:13.773576 8667810944 trainers.py:102] Trainer restoring the full model\n",
      "I0518 15:02:13.800570 8667810944 trainers.py:123] Loaded checkpoint ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/ckpt-16500\n",
      "I0518 15:02:13.800647 8667810944 trainers.py:124] Loading model took 0.0 seconds\n",
      "I0518 15:02:18.905035 8667810944 train_util.py:282] Creating metrics for ['spectral_loss', 'total_loss']\n",
      "I0518 15:02:18.915752 8667810944 train_util.py:295] step: 16501\tspectral_loss: 6.70\ttotal_loss: 6.70\t\n",
      "I0518 15:02:19.995949 8667810944 train_util.py:295] step: 16502\tspectral_loss: 6.55\ttotal_loss: 6.55\t\n",
      "I0518 15:02:20.949423 8667810944 train_util.py:295] step: 16503\tspectral_loss: 6.34\ttotal_loss: 6.34\t\n",
      "I0518 15:02:21.872890 8667810944 train_util.py:295] step: 16504\tspectral_loss: 6.35\ttotal_loss: 6.35\t\n",
      "I0518 15:02:22.785866 8667810944 train_util.py:295] step: 16505\tspectral_loss: 6.04\ttotal_loss: 6.04\t\n",
      "I0518 15:02:23.734056 8667810944 train_util.py:295] step: 16506\tspectral_loss: 6.23\ttotal_loss: 6.23\t\n",
      "I0518 15:02:24.664860 8667810944 train_util.py:295] step: 16507\tspectral_loss: 5.81\ttotal_loss: 5.81\t\n",
      "I0518 15:02:25.573765 8667810944 train_util.py:295] step: 16508\tspectral_loss: 6.44\ttotal_loss: 6.44\t\n",
      "I0518 15:02:26.500550 8667810944 train_util.py:295] step: 16509\tspectral_loss: 6.51\ttotal_loss: 6.51\t\n",
      "I0518 15:02:27.470898 8667810944 train_util.py:295] step: 16510\tspectral_loss: 6.23\ttotal_loss: 6.23\t\n",
      "I0518 15:02:28.491012 8667810944 train_util.py:295] step: 16511\tspectral_loss: 6.60\ttotal_loss: 6.60\t\n",
      "I0518 15:02:29.390116 8667810944 train_util.py:295] step: 16512\tspectral_loss: 6.89\ttotal_loss: 6.89\t\n",
      "I0518 15:02:30.268237 8667810944 train_util.py:295] step: 16513\tspectral_loss: 7.00\ttotal_loss: 7.00\t\n",
      "I0518 15:02:31.143017 8667810944 train_util.py:295] step: 16514\tspectral_loss: 7.40\ttotal_loss: 7.40\t\n",
      "I0518 15:02:32.021173 8667810944 train_util.py:295] step: 16515\tspectral_loss: 6.88\ttotal_loss: 6.88\t\n",
      "I0518 15:02:32.905416 8667810944 train_util.py:295] step: 16516\tspectral_loss: 7.10\ttotal_loss: 7.10\t\n",
      "I0518 15:02:33.776572 8667810944 train_util.py:295] step: 16517\tspectral_loss: 7.67\ttotal_loss: 7.67\t\n",
      "I0518 15:02:34.648252 8667810944 train_util.py:295] step: 16518\tspectral_loss: 7.50\ttotal_loss: 7.50\t\n",
      "I0518 15:02:35.525063 8667810944 train_util.py:295] step: 16519\tspectral_loss: 7.76\ttotal_loss: 7.76\t\n",
      "I0518 15:02:36.397485 8667810944 train_util.py:295] step: 16520\tspectral_loss: 8.05\ttotal_loss: 8.05\t\n",
      "I0518 15:02:37.270870 8667810944 train_util.py:295] step: 16521\tspectral_loss: 7.50\ttotal_loss: 7.50\t\n",
      "I0518 15:02:38.143230 8667810944 train_util.py:295] step: 16522\tspectral_loss: 8.34\ttotal_loss: 8.34\t\n",
      "I0518 15:02:39.038110 8667810944 train_util.py:295] step: 16523\tspectral_loss: 7.99\ttotal_loss: 7.99\t\n",
      "I0518 15:02:39.931425 8667810944 train_util.py:295] step: 16524\tspectral_loss: 7.70\ttotal_loss: 7.70\t\n",
      "I0518 15:02:40.825921 8667810944 train_util.py:295] step: 16525\tspectral_loss: 7.38\ttotal_loss: 7.38\t\n",
      "I0518 15:02:41.702409 8667810944 train_util.py:295] step: 16526\tspectral_loss: 7.69\ttotal_loss: 7.69\t\n",
      "I0518 15:02:42.572983 8667810944 train_util.py:295] step: 16527\tspectral_loss: 7.29\ttotal_loss: 7.29\t\n",
      "I0518 15:02:43.457271 8667810944 train_util.py:295] step: 16528\tspectral_loss: 8.28\ttotal_loss: 8.28\t\n",
      "I0518 15:02:44.332221 8667810944 train_util.py:295] step: 16529\tspectral_loss: 7.46\ttotal_loss: 7.46\t\n",
      "I0518 15:02:45.203021 8667810944 train_util.py:295] step: 16530\tspectral_loss: 7.84\ttotal_loss: 7.84\t\n",
      "I0518 15:02:46.078284 8667810944 train_util.py:295] step: 16531\tspectral_loss: 8.20\ttotal_loss: 8.20\t\n",
      "I0518 15:02:46.946815 8667810944 train_util.py:295] step: 16532\tspectral_loss: 8.31\ttotal_loss: 8.31\t\n",
      "I0518 15:02:47.814746 8667810944 train_util.py:295] step: 16533\tspectral_loss: 6.25\ttotal_loss: 6.25\t\n",
      "I0518 15:02:48.676891 8667810944 train_util.py:295] step: 16534\tspectral_loss: 5.39\ttotal_loss: 5.39\t\n",
      "I0518 15:02:49.552757 8667810944 train_util.py:295] step: 16535\tspectral_loss: 5.71\ttotal_loss: 5.71\t\n",
      "I0518 15:02:50.419560 8667810944 train_util.py:295] step: 16536\tspectral_loss: 5.93\ttotal_loss: 5.93\t\n",
      "I0518 15:02:51.296170 8667810944 train_util.py:295] step: 16537\tspectral_loss: 5.92\ttotal_loss: 5.92\t\n",
      "I0518 15:02:52.165183 8667810944 train_util.py:295] step: 16538\tspectral_loss: 5.91\ttotal_loss: 5.91\t\n",
      "I0518 15:02:53.056746 8667810944 train_util.py:295] step: 16539\tspectral_loss: 7.01\ttotal_loss: 7.01\t\n",
      "I0518 15:02:53.918138 8667810944 train_util.py:295] step: 16540\tspectral_loss: 7.62\ttotal_loss: 7.62\t\n",
      "I0518 15:02:54.801877 8667810944 train_util.py:295] step: 16541\tspectral_loss: 7.04\ttotal_loss: 7.04\t\n",
      "I0518 15:02:55.668406 8667810944 train_util.py:295] step: 16542\tspectral_loss: 7.10\ttotal_loss: 7.10\t\n",
      "I0518 15:02:56.544547 8667810944 train_util.py:295] step: 16543\tspectral_loss: 6.67\ttotal_loss: 6.67\t\n",
      "I0518 15:02:57.425894 8667810944 train_util.py:295] step: 16544\tspectral_loss: 6.19\ttotal_loss: 6.19\t\n",
      "I0518 15:02:58.303447 8667810944 train_util.py:295] step: 16545\tspectral_loss: 5.77\ttotal_loss: 5.77\t\n",
      "I0518 15:02:59.179970 8667810944 train_util.py:295] step: 16546\tspectral_loss: 6.99\ttotal_loss: 6.99\t\n",
      "I0518 15:03:00.092228 8667810944 train_util.py:295] step: 16547\tspectral_loss: 6.94\ttotal_loss: 6.94\t\n",
      "I0518 15:03:00.989335 8667810944 train_util.py:295] step: 16548\tspectral_loss: 5.57\ttotal_loss: 5.57\t\n",
      "I0518 15:03:01.873920 8667810944 train_util.py:295] step: 16549\tspectral_loss: 7.29\ttotal_loss: 7.29\t\n",
      "I0518 15:03:02.760093 8667810944 train_util.py:295] step: 16550\tspectral_loss: 7.71\ttotal_loss: 7.71\t\n",
      "I0518 15:03:03.659609 8667810944 train_util.py:295] step: 16551\tspectral_loss: 7.25\ttotal_loss: 7.25\t\n",
      "I0518 15:03:04.549434 8667810944 train_util.py:295] step: 16552\tspectral_loss: 4.67\ttotal_loss: 4.67\t\n",
      "I0518 15:03:05.451770 8667810944 train_util.py:295] step: 16553\tspectral_loss: 5.42\ttotal_loss: 5.42\t\n",
      "I0518 15:03:06.342830 8667810944 train_util.py:295] step: 16554\tspectral_loss: 5.53\ttotal_loss: 5.53\t\n",
      "I0518 15:03:07.235484 8667810944 train_util.py:295] step: 16555\tspectral_loss: 5.43\ttotal_loss: 5.43\t\n",
      "I0518 15:03:08.116955 8667810944 train_util.py:295] step: 16556\tspectral_loss: 5.46\ttotal_loss: 5.46\t\n",
      "I0518 15:03:09.015510 8667810944 train_util.py:295] step: 16557\tspectral_loss: 5.49\ttotal_loss: 5.49\t\n",
      "I0518 15:03:09.896519 8667810944 train_util.py:295] step: 16558\tspectral_loss: 5.41\ttotal_loss: 5.41\t\n",
      "I0518 15:03:10.773246 8667810944 train_util.py:295] step: 16559\tspectral_loss: 5.34\ttotal_loss: 5.34\t\n",
      "I0518 15:03:11.662298 8667810944 train_util.py:295] step: 16560\tspectral_loss: 5.34\ttotal_loss: 5.34\t\n",
      "I0518 15:03:12.537117 8667810944 train_util.py:295] step: 16561\tspectral_loss: 5.58\ttotal_loss: 5.58\t\n",
      "I0518 15:03:13.421664 8667810944 train_util.py:295] step: 16562\tspectral_loss: 5.53\ttotal_loss: 5.53\t\n",
      "I0518 15:03:14.298198 8667810944 train_util.py:295] step: 16563\tspectral_loss: 5.61\ttotal_loss: 5.61\t\n",
      "I0518 15:03:15.166302 8667810944 train_util.py:295] step: 16564\tspectral_loss: 5.33\ttotal_loss: 5.33\t\n",
      "I0518 15:03:16.034410 8667810944 train_util.py:295] step: 16565\tspectral_loss: 5.26\ttotal_loss: 5.26\t\n",
      "I0518 15:03:16.892902 8667810944 train_util.py:295] step: 16566\tspectral_loss: 5.30\ttotal_loss: 5.30\t\n",
      "I0518 15:03:17.763159 8667810944 train_util.py:295] step: 16567\tspectral_loss: 5.28\ttotal_loss: 5.28\t\n",
      "I0518 15:03:18.634773 8667810944 train_util.py:295] step: 16568\tspectral_loss: 5.44\ttotal_loss: 5.44\t\n",
      "I0518 15:03:19.507291 8667810944 train_util.py:295] step: 16569\tspectral_loss: 5.37\ttotal_loss: 5.37\t\n",
      "I0518 15:03:20.419141 8667810944 train_util.py:295] step: 16570\tspectral_loss: 5.52\ttotal_loss: 5.52\t\n",
      "I0518 15:03:21.295773 8667810944 train_util.py:295] step: 16571\tspectral_loss: 5.30\ttotal_loss: 5.30\t\n",
      "I0518 15:03:22.161684 8667810944 train_util.py:295] step: 16572\tspectral_loss: 5.37\ttotal_loss: 5.37\t\n",
      "I0518 15:03:23.047710 8667810944 train_util.py:295] step: 16573\tspectral_loss: 5.37\ttotal_loss: 5.37\t\n",
      "I0518 15:03:23.941614 8667810944 train_util.py:295] step: 16574\tspectral_loss: 5.32\ttotal_loss: 5.32\t\n",
      "I0518 15:03:24.810682 8667810944 train_util.py:295] step: 16575\tspectral_loss: 5.29\ttotal_loss: 5.29\t\n",
      "I0518 15:03:25.676383 8667810944 train_util.py:295] step: 16576\tspectral_loss: 5.15\ttotal_loss: 5.15\t\n",
      "I0518 15:03:26.557385 8667810944 train_util.py:295] step: 16577\tspectral_loss: 4.93\ttotal_loss: 4.93\t\n",
      "I0518 15:03:27.440146 8667810944 train_util.py:295] step: 16578\tspectral_loss: 6.04\ttotal_loss: 6.04\t\n",
      "I0518 15:03:28.334455 8667810944 train_util.py:295] step: 16579\tspectral_loss: 6.17\ttotal_loss: 6.17\t\n",
      "I0518 15:03:29.222618 8667810944 train_util.py:295] step: 16580\tspectral_loss: 6.16\ttotal_loss: 6.16\t\n",
      "I0518 15:03:30.091372 8667810944 train_util.py:295] step: 16581\tspectral_loss: 6.19\ttotal_loss: 6.19\t\n",
      "I0518 15:03:30.980740 8667810944 train_util.py:295] step: 16582\tspectral_loss: 6.04\ttotal_loss: 6.04\t\n",
      "I0518 15:03:31.899471 8667810944 train_util.py:295] step: 16583\tspectral_loss: 6.00\ttotal_loss: 6.00\t\n",
      "I0518 15:03:32.841813 8667810944 train_util.py:295] step: 16584\tspectral_loss: 5.93\ttotal_loss: 5.93\t\n",
      "I0518 15:03:33.725057 8667810944 train_util.py:295] step: 16585\tspectral_loss: 5.99\ttotal_loss: 5.99\t\n",
      "I0518 15:03:34.629755 8667810944 train_util.py:295] step: 16586\tspectral_loss: 6.62\ttotal_loss: 6.62\t\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def run_training(drive_dir=''):\n",
    "  log_event('runTrainingStarted', {})\n",
    "\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # Setup\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # Save data locally, but model on drive.\n",
    "  data_dir = 'data/'\n",
    "  audio_dir = 'audio/'\n",
    "  model_dir = get_model_dir(drive_dir)\n",
    "\n",
    "  # NOTE: do not reset!\n",
    "  # reset_state(data_dir, audio_dir, model_dir)\n",
    "\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # Dataset\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # tick = time.time()\n",
    "\n",
    "  # get_audio_files(drive_dir, audio_dir)\n",
    "  # prepare_dataset(audio_dir, data_dir)\n",
    "\n",
    "  # log_event('datasetMins', {'value': round((time.time() - tick) // 60)})\n",
    "\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # Train\n",
    "  # ------------------------------------------------------------------------------\n",
    "  tick = time.time()\n",
    "\n",
    "  print()\n",
    "  print('Training...')\n",
    "  train(model_dir, data_dir, steps=Training_Steps)\n",
    "\n",
    "  log_event('trainMins', {\n",
    "      'event_category': str(Training_Steps),\n",
    "      'value': round((time.time() - tick) // 60),\n",
    "  })\n",
    "\n",
    "  # ------------------------------------------------------------------------------\n",
    "  # Export\n",
    "  # ------------------------------------------------------------------------------\n",
    "  tick = time.time()\n",
    "\n",
    "  print()\n",
    "  print('Exporting model...')\n",
    "  export_and_download(model_dir)\n",
    "\n",
    "  log_event('exportMins', {'value': round((time.time() - tick) // 60)})\n",
    "\n",
    "\n",
    "def run(Google_Drive=True):\n",
    "  \"\"\"Create and display a FileChooser widget.\"\"\"\n",
    "  log_event('runStarted', {})\n",
    "  gpu_type = 'mac'\n",
    "  log_event('gpuType', {'event_category': gpu_type})\n",
    "  log_event('trainingLocally', {})\n",
    "  run_training(drive_dir=Sample_Dir)\n",
    "    \n",
    "# The single command.\n",
    "Google_Drive = False\n",
    "run(Google_Drive)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f2e06c25-4929-48c8-a209-c98b3c29bfe8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing from operative_config ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/operative_config-16500.gin\n",
      "WARNING:tensorflow:Layer gru will not use cuDNN kernels since it doesn't meet the criteria. It will use a generic GPU kernel as fallback when running on GPU.\n",
      "W0518 17:00:59.082592 8667810944 gru.py:618] Layer gru will not use cuDNN kernels since it doesn't meet the criteria. It will use a generic GPU kernel as fallback when running on GPU.\n",
      "I0518 17:00:59.089889 8667810944 model.py:90] Model restoring all components.\n",
      "I0518 17:00:59.094999 8667810944 model.py:109] Loaded checkpoint ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/ckpt-24300\n",
      "I0518 17:00:59.095035 8667810944 model.py:110] Loading model took 0.0 seconds\n",
      "Reverb Impulse Response is 1.5 seconds long\n",
      "Parsing from operative_config ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/operative_config-16500.gin\n",
      "Parsing from operative_config ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/operative_config-16500.gin\n",
      "Calculating dataset statistics for <ddsp.training.data.TFRecordProvider object at 0x315c87160>\n",
      "Computing statistics for 1241 examples.\n",
      "\n",
      "Converting to SavedModel:\n",
      "Input: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917\n",
      "Output: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum\n",
      "\n",
      "Parsing --gin_param flags: []\n",
      "Parsing from operative_config ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/operative_config-16500.gin\n",
      "WARNING:tensorflow:Layer gru_1 will not use cuDNN kernels since it doesn't meet the criteria. It will use a generic GPU kernel as fallback when running on GPU.\n",
      "W0518 17:01:12.746224 8667810944 gru.py:618] Layer gru_1 will not use cuDNN kernels since it doesn't meet the criteria. It will use a generic GPU kernel as fallback when running on GPU.\n",
      "I0518 17:01:12.753224 8667810944 model.py:99] Trainer restoring model subcomponents:\n",
      "I0518 17:01:12.753262 8667810944 model.py:103] Restoring {'decoder': <ddsp.training.decoders.RnnFcDecoder object at 0x3140bf520>}\n",
      "I0518 17:01:12.756944 8667810944 model.py:109] Loaded checkpoint ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/ckpt-24300\n",
      "I0518 17:01:12.757249 8667810944 model.py:110] Loading model took 0.0 seconds\n",
      "Inputs to Model: ([1], [1], [512])\n",
      "WARNING:tensorflow:Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
      "W0518 17:01:13.348746 8667810944 checkpoint.py:208] Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._iterations\n",
      "W0518 17:01:13.348833 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._iterations\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._current_learning_rate\n",
      "W0518 17:01:13.348851 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._current_learning_rate\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
      "W0518 17:01:13.348865 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
      "W0518 17:01:13.349006 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
      "W0518 17:01:13.349027 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
      "W0518 17:01:13.349040 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
      "W0518 17:01:13.349050 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
      "W0518 17:01:13.349129 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
      "W0518 17:01:13.349143 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
      "W0518 17:01:13.349153 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
      "W0518 17:01:13.349164 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
      "W0518 17:01:13.349265 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
      "W0518 17:01:13.349280 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
      "W0518 17:01:13.349291 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
      "W0518 17:01:13.349302 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
      "W0518 17:01:13.349406 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
      "W0518 17:01:13.349419 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
      "W0518 17:01:13.349430 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
      "W0518 17:01:13.349441 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n",
      "W0518 17:01:13.349545 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.19\n",
      "W0518 17:01:13.349560 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.19\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.20\n",
      "W0518 17:01:13.349570 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.20\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.21\n",
      "W0518 17:01:13.349676 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.21\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.22\n",
      "W0518 17:01:13.349690 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.22\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.23\n",
      "W0518 17:01:13.349702 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.23\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.24\n",
      "W0518 17:01:13.349712 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.24\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.25\n",
      "W0518 17:01:13.349808 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.25\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.26\n",
      "W0518 17:01:13.349823 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.26\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.27\n",
      "W0518 17:01:13.349834 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.27\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.28\n",
      "W0518 17:01:13.349846 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.28\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.29\n",
      "W0518 17:01:13.349951 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.29\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.30\n",
      "W0518 17:01:13.349966 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.30\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.31\n",
      "W0518 17:01:13.349978 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.31\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.32\n",
      "W0518 17:01:13.349989 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.32\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.33\n",
      "W0518 17:01:13.350085 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.33\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.34\n",
      "W0518 17:01:13.350101 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.34\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.35\n",
      "W0518 17:01:13.350112 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.35\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.36\n",
      "W0518 17:01:13.350124 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.36\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.dense_out.kernel\n",
      "W0518 17:01:13.350223 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.dense_out.kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.dense_out.bias\n",
      "W0518 17:01:13.350237 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.dense_out.bias\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.kernel\n",
      "W0518 17:01:13.350250 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.recurrent_kernel\n",
      "W0518 17:01:13.350260 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.recurrent_kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.bias\n",
      "W0518 17:01:13.350363 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.rnn.rnn.cell.bias\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "W0518 17:01:13.350376 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-0.bias\n",
      "W0518 17:01:13.350389 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-0.bias\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "W0518 17:01:13.350494 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-1.beta\n",
      "W0518 17:01:13.350511 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.out_stack.layer_with_weights-0.layer_with_weights-1.beta\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "W0518 17:01:13.350522 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-0.bias\n",
      "W0518 17:01:13.350639 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-0.bias\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "W0518 17:01:13.350651 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-1.beta\n",
      "W0518 17:01:13.350662 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.0.layer_with_weights-0.layer_with_weights-1.beta\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "W0518 17:01:13.350773 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-0.kernel\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-0.bias\n",
      "W0518 17:01:13.350785 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-0.bias\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "W0518 17:01:13.350796 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-1.gamma\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-1.beta\n",
      "W0518 17:01:13.350905 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.decoder.input_stacks.1.layer_with_weights-0.layer_with_weights-1.beta\n",
      "Outputs from Model: ([1], [60], [65], [512])\n",
      "Finshed Loading Model!\n",
      "I0518 17:01:13.733130 8667810944 signature_serialization.py:148] Function `call` contains input name(s) resource with unsupported characters which will be renamed to rnn_fc_decoder_1_dense_7_biasadd_readvariableop_resource in the SavedModel.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.training.preprocessing.OnlineF0PowerPreprocessor object at 0x314320e50>, because it is not built.\n",
      "W0518 17:01:13.883340 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.training.preprocessing.OnlineF0PowerPreprocessor object at 0x314320e50>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.processors.ProcessorGroup object at 0x3170bbb50>, because it is not built.\n",
      "W0518 17:01:14.109200 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.processors.ProcessorGroup object at 0x3170bbb50>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.losses.SpectralLoss object at 0x3170b6c70>, because it is not built.\n",
      "W0518 17:01:14.109274 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.losses.SpectralLoss object at 0x3170b6c70>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.synths.Harmonic object at 0x314197280>, because it is not built.\n",
      "W0518 17:01:14.779894 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.synths.Harmonic object at 0x314197280>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.synths.FilteredNoise object at 0x314158bb0>, because it is not built.\n",
      "W0518 17:01:14.780507 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.synths.FilteredNoise object at 0x314158bb0>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.processors.Add object at 0x3141f27f0>, because it is not built.\n",
      "W0518 17:01:14.781062 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.processors.Add object at 0x3141f27f0>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.effects.FilteredNoiseReverb object at 0x3170d1310>, because it is not built.\n",
      "W0518 17:01:14.781594 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.effects.FilteredNoiseReverb object at 0x3170d1310>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.processors.Crop object at 0x3170d1520>, because it is not built.\n",
      "W0518 17:01:14.782245 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.processors.Crop object at 0x3170d1520>, because it is not built.\n",
      "WARNING:tensorflow:Skipping full serialization of Keras layer <ddsp.synths.FilteredNoise object at 0x3170d15e0>, because it is not built.\n",
      "W0518 17:01:14.792725 8667810944 save_impl.py:66] Skipping full serialization of Keras layer <ddsp.synths.FilteredNoise object at 0x3170d15e0>, because it is not built.\n",
      "I0518 17:01:14.941976 8667810944 save.py:274] Found untraced functions such as stateless_rnn_layer_call_fn, stateless_rnn_layer_call_and_return_conditional_losses, dense_7_layer_call_fn, dense_7_layer_call_and_return_conditional_losses while saving (showing 4 of 4). These functions will not be directly callable after loading.\n",
      "INFO:tensorflow:Assets written to: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum/assets\n",
      "I0518 17:01:16.073982 8667810944 builder_impl.py:804] Assets written to: ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum/assets\n",
      "I0518 17:01:16.512890 8667810944 fingerprinting_utils.py:48] Writing fingerprint to ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum/fingerprint.pb\n",
      "SavedModel Conversion Success!\n",
      "\n",
      "Converting to TFLite:\n",
      "Input:./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum\n",
      "Output:./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum/tflite\n",
      "\n",
      "WARNING:tensorflow:Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
      "W0518 17:01:16.892673 8667810944 checkpoint.py:208] Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._iterations\n",
      "W0518 17:01:16.892755 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._iterations\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._current_learning_rate\n",
      "W0518 17:01:16.892773 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._current_learning_rate\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
      "W0518 17:01:16.892786 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
      "W0518 17:01:16.892922 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
      "W0518 17:01:16.892943 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
      "W0518 17:01:16.892956 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
      "W0518 17:01:16.892966 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
      "W0518 17:01:16.893045 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
      "W0518 17:01:16.893060 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
      "W0518 17:01:16.893069 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
      "W0518 17:01:16.893081 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
      "W0518 17:01:16.893183 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
      "W0518 17:01:16.893197 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
      "W0518 17:01:16.893208 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
      "W0518 17:01:16.893218 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
      "W0518 17:01:16.893317 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
      "W0518 17:01:16.893331 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
      "W0518 17:01:16.893342 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
      "W0518 17:01:16.893352 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n",
      "W0518 17:01:16.893749 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.19\n",
      "W0518 17:01:16.893765 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.19\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.20\n",
      "W0518 17:01:16.893777 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.20\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.21\n",
      "W0518 17:01:16.893902 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.21\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.22\n",
      "W0518 17:01:16.893915 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.22\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.23\n",
      "W0518 17:01:16.893926 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.23\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.24\n",
      "W0518 17:01:16.893937 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.24\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.25\n",
      "W0518 17:01:16.894054 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.25\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.26\n",
      "W0518 17:01:16.894075 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.26\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.27\n",
      "W0518 17:01:16.894090 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.27\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.28\n",
      "W0518 17:01:16.894105 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.28\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.29\n",
      "W0518 17:01:16.894189 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.29\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.30\n",
      "W0518 17:01:16.894204 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.30\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.31\n",
      "W0518 17:01:16.894217 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.31\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.32\n",
      "W0518 17:01:16.894230 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.32\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.33\n",
      "W0518 17:01:16.894330 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.33\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.34\n",
      "W0518 17:01:16.894345 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.34\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.35\n",
      "W0518 17:01:16.894357 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.35\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.36\n",
      "W0518 17:01:16.894369 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).optimizer._variables.36\n",
      "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).model.processor_group.reverb.magnitudes\n",
      "W0518 17:01:16.894474 8667810944 checkpoint.py:217] Value in checkpoint could not be found in the restored object: (root).model.processor_group.reverb.magnitudes\n",
      "2025-05-18 17:01:19.538086: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:364] Ignored output_format.\n",
      "2025-05-18 17:01:19.538110: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:367] Ignored drop_control_dependency.\n",
      "/Users/yilinzhang/Documents/coding/ddsp_proj/.venv/lib/python3.8/site-packages/tflite_support/metadata.py:273: UserWarning: File, 'metadata.json', does not exsit in the metadata. But packing it to tflite model is still allowed.\n",
      "  warnings.warn(\n",
      "TFLite Conversion Success!\n",
      "Export complete! Zipping ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum to ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum.zip\n",
      "updating: gayageum/ (stored 0%)\n",
      "updating: gayageum/reverb_ir_44100_hz.wav (deflated 48%)\n",
      "updating: gayageum/reverb_ir_16000_hz.wav (deflated 48%)\n",
      "updating: gayageum/tflite/ (stored 0%)\n",
      "updating: gayageum/tflite/gayageum.tflite (deflated 8%)\n",
      "updating: gayageum/fingerprint.pb (stored 0%)\n",
      "updating: gayageum/metadata.json (deflated 45%)\n",
      "updating: gayageum/variables/ (stored 0%)\n",
      "updating: gayageum/variables/variables.data-00000-of-00001 (deflated 8%)\n",
      "updating: gayageum/variables/variables.index (deflated 59%)\n",
      "updating: gayageum/saved_model.pb (deflated 89%)\n",
      "updating: gayageum/assets/ (stored 0%)\n",
      "updating: gayageum/.DS_Store (deflated 96%)\n",
      "Zipping Complete! Downloading... gayageum.zip\n",
      "You can also find your model at ./ddsp_train/gayageum_samples_all/ddsp-training-2025-05-18-0917/gayageum\n"
     ]
    }
   ],
   "source": [
    "export_and_download(get_model_dir(Sample_Dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "74358f6a-ac37-497b-98e8-1f9fee9c8151",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a2d27a18-4e71-46e2-a827-555fd580cbb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "\n",
    "tf.config.list_physical_devices(\"GPU\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a917b1d6-84d1-4369-88c6-73fdb5c9fe1a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
